### ISSUE TITLE
Add `emit.warning()`

### ISSUE THREAD

**Issue by mr-cal** [canonical/craft-cli#256]
### What needs to get done

Add a new emitter command, `emit.warning()`

### Why it needs to get done

`emit.progress(permanent=True)` is overloaded.  It is the only way to log warnings like deprecation notices.

#### Acceptance criteria

* Warning method on the emitter that outputs (permanently) to stderr, includes the warning in the log, and provides a customisable prefix (Default: “WARNING: ”)  
* Tests for this method  
* At least one example in [examples.py](http://examples.py) that uses progress with permanent=False, progress with permanent=True, and warning.

**Comment by beliaev-maksim** [comment]
### **💰 Bounty – $300 USD**
Bounty is sponsored by **Canonical**, the company behind Ubuntu.


### **📜 Rules:**

1. You must be enrolled in the [**GitHub Sponsors**](https://docs.github.com/en/sponsors/getting-started-with-github-sponsors/about-github-sponsors) program.  
2. Contribution rules from this repository apply. Please carefully read CONTRIBUTING.md  
3. The bounty will be paid **when the implementation meets the Business and Technical Specifications**.  
4. The bounty will be paid **when the solution is merged into the `main` or `release` branch**.  
5. Contributors working on the issue **must be assigned by the repository maintainers**.  
6. The bounty will be awarded to the **first accepted solution**.
7. To keep things fair, each contributor can work on only one bounty at a time. Once it’s completed, you’re free to pick the next challenge :)

💡 **Note:** The bounty amount is **gross**. Any applicable **fees, taxes, or deductions** from payment providers (e.g., GitHub Sponsors, bank transfers) may apply, and **Canonical is not responsible for covering these additional costs**. 

Happy hacking\! 🚀

For any questions and payouts please reach out to [community+bounty@canonical.com](mailto:community+bounty@canonical.com)

<a href="https://canonical.com/">
  <img src="https://assets.ubuntu.com/v1/36651171-Canonical-Logo-Light_bkg@2x.png" height="100" style="background:white; padding:10px; border-radius:5px;">
</a>

**Comment by nryanl** [comment]
Hello, I would like to work on this!

Here is my approach:

## Approach

1) Write `emit.warning()` 
- Supports a customizable prefix (provided via kwarg)
- Is permanent and displays regardless of `EmitterMode`
- Follows the example of existing emit methods

2) Write `assert_warning(self, expected_text, regex=False)` in `craft_cli/pytest_plugin.py`
- Simply `return self._check(expected_text, "warning", regex)`

3) Add both integration and unit tests for `emit.warning()`, following the example of existing tests for the emitter message methods.

4) Add examples to showcase the differences between `progress()` (with permanent=False/True) and `warning()`.

**Comment by Mr-Sunglasses** [comment]
Hi I would like to work on this issue!

Here is my high-level solution plan:

-------

1. Add a `warning()` Method in `Emitter` class.

-  it provides dedicated warning message functionality

- method Signature
```
def warning(self, text: str, prefix: str = "WARNING: ") -> None
```

- method specs.

- [x] Always outputs to stderr with permanent visibility
- [x] Includes all warnings in log files
- [x] Default "WARNING: " with override capability
- [x] Respects emitter mode conventions

- Compatibility
   - Existing `progress(permanent=True)` usage remains functional
   - Logging: Integrate with existing log infrastructure

-------

2. Test Coverage
   - Coverage Areas:
     - All emitter modes (QUIET, BRIEF, VERBOSE, DEBUG, TRACE)
     - Custom prefix functionality
     - Empty prefix edge case
     - Integration with existing emitter patterns

-------

3. Usage Examples

Demonstrates:
- Ephemeral progress (permanent=False)
- Permanent progress (permanent=True)
- Warning messages with various prefixes

---------

I've also create a draft PR of my implementation #355

**Comment by sergiusens** [comment]
@nryanl we are assigning this to you

**Comment by beliaev-maksim** [comment]
@nryanl @sergiusens 
I am scheduling the payout for the next Monday to process

**Comment by beliaev-maksim** [comment]
@nryanl Update: we are currently processing the invoice, and once the funds land on GitHub, we will pay the bounty out. We are doing it as soon as possible, but please allow 5–10 business days.

cc: @sergiusens

**Comment by beliaev-maksim** [comment]
@nryanl the bounty was paid and should be on your account soon
